### PR TITLE
Refactor the way that `Button` components are styled

### PR DESCRIPTION
**Pull request by lyzadanger**
This draft PR shows an implementation approach for, with luck, _stopping the insanity_ that is trying to style `<button>` elements generated by the `Button` component.

Before discussing implementation motivations, let's talk about **user-visible changes** in this changeset.

### Discernible "pressed", "expanded" (On) states

This was actually the initial motivation that spurred these changes. We've been discussing one design variant of moving the search field off of the top bar, but perhaps leaving the search icon toggle there in the top bar. To do this successfully, I think we need to have visible "on" states for buttons and menu toggles. That has been accomplished in this PR:

image

image

### Fixing some pre-existing responsive/narrow-viewport bugs

Lots of the more recent button styling was done without checking what things look like in small viewports. This work makes a litany of small changes to address button styling in narrow viewports. By no means do I claim that things are perfect now, but they are _better_.

### Small consistency tweaks

We have several "link-style" buttons that are styled to look like they're inline text, or links. The colors of these, and their `text-decoration` have been consolidated.

e.g. This button now has an underline to better match its friends (you're seeing the hover state here):

image

_However_, for the most part, this PR attempts to make minimal changes; it's goals are to refactor/reorganize.

## Implementation approach/motivation

**tl;dr** *The `Button` component will still apply some basic default styles, but those can be overridden by setting a `className` prop. There are mixins defined for each general button design pattern to make "extending" (i.e. overriding) styles more straightforward when a `className` is provided).*

Previously, the `Button` component was trying to be pretty clever about applying a pastiche of classes to account for several button "types" and states. Some of these were inferred internally by the component, while some were communicated via props (`useCompactStyle`, e.g.).

A `className` prop existed before this set of changes, but was used _additively_ with the component's own classes. This led to pain with CSS specificity when it was desirable to override the component's default style rules. It also made it hard to see what was going on.

On the other hand, it's useful to have a component that handles rendering an appropriate icon and applying consistent markup and `aria-*` attributes.

I've made an attempt to ferret out some of the more Byzantine styling duties of the component and simplify its default styles. As of these changes, the component will always apply exactly one className to the `button` element, and one className to the `icon` it renders (if any) within that button.

The button classnames will be one of three combinations as of these changes:

1. When there is no text/label displayed, only an icon (and no `className` prop is present)

    ```
    <button className="button--icon-only">
        <SvgIcon ... className="button--icon-only__icon" />
    </button>
    ```

2. Where there is text/label displayed, optionally an icon (and no `className` prop is present):

    ```
    <button className="button--labeled">
        <SvgIcon ... className="button--labeled__icon />
    </button>
    ```

3. When `className` is present:

    ```
    <button className={className}>
        <SvgIcon ... className={`${className}__icon`} />
    </button>
    ```

That's one prong of the work. The other prong is reorganizing the SASS mixins and CSS classes to make easier-to-use, reusable classes and mixins for buttons that want to start with base styles but extend/overrride some of them. The button mixins defined in this branch are intended to start coalescing (descriptively, not prescriptively nor aspirationally) some of our core button design patterns.

The default `.button-*` classes make direct use of some of those mixins:

```
.button {
  @include buttons.button;
}

.button--icon-only {
  @include buttons.button--icon-only;
}

.button--labeled {
  @include buttons.button--labeled;
}

```

While the full button mixin list as of now contains some additional items:

* `button--primary`: Builds on `button--labeled` to make a labeled button into a primary action button
* `button--input`: A pattern used when an icon-only button is displayed adjacent to a text input field (see annotation-share controls for example)
* `button--link`: We have a scattering of buttons that aren't really meant to look like buttons but instead look like "regular text" or a link.

## What's not done in this work

* This is a draft PR/work. Tests are going to fail.
* I haven't converted a single `<button>` to a `Button`. That's work for later once the approach is sorted out.
* Other than the styling of "textlike" buttons mentioned above, and the "on" states, I've tried to resist the temptation to fix and consolidate inconsistent button design patterns: font, colors, layout, etc.

## Some High-level Thoughts

This work feels like high-level thinking because we haven't fully established how to make our UI framework and SASS and design patterns work together for reusable atomic components like `button`s. Some of this is an exploration of how to encapsulate those patterns from an implementation perspective.

Overall, these changes are primarily _reductive_ in complexity: `Button` has fewer props, and the known button design patterns are distilled into the `buttons` SASS module, with fewer component-CSS-specific rules.

I wanted to get something in a draft way here before being unavailable next week...

### CODE REVIEW

File: src/sidebar/components/annotation-share-control.js [hunk 1]
```diff
@@ -87,7 +87,12 @@ function AnnotationShareControl({
 
   return (
     <div className="annotation-share-control" ref={shareRef}>
-      <Button icon="share" title="Share" onClick={toggleSharePanel} />
+      <Button
+        icon="share"
+        title="Share"
+        onClick={toggleSharePanel}
+        isExpanded={isOpen}
```
Comment (lyzadanger) [inline src/sidebar/components/annotation-share-control.js:94]: This was a missing prop that should have been here before (a11y-wise).

File: src/styles/mixins/buttons.scss [hunk 2]
```diff
@@ -1,6 +1,8 @@
 @use "./focus";
 @use "../variables" as var;
 
+$button-padding: 0.5em;
```
Comment (lyzadanger) [inline src/styles/mixins/buttons.scss:4]: This is only used once and is probably not useful; artifact of an earlier approach.

File: src/styles/sidebar/components/annotation-body.scss [hunk 3]
```diff
@@ -38,6 +39,7 @@
   justify-content: flex-end;
 
   .annotation-body__collapse-toggle-button {
+    @include buttons.button--labeled;
```
Comment (lyzadanger) [inline src/styles/sidebar/components/annotation-body.scss:42]: So, as a general example here: `Annotation` can pass `annotation-body__collapse-toggle-button` as a `className` to a `Button` and then do whatever overrides to button styling (based on `buttons.button--labeled`) it wants to here.

File: src/styles/sidebar/components/button.scss [hunk 4]
```diff
@@ -2,73 +2,23 @@
...
+ * default styles are not overridden:
  *
- * <button class="button <button--icon-only|button--labeled> [button--compact] [button--primary] [is-active]">
- *   [<SvgIcon class="button__icon" />]
+ * <button class="<button--icon-only|button--labeled>">
+ *   [<SvgIcon class="<button--icon-only__icon|button--labeled__icon>" />]
  *   [<span>label</span>]
  * </button>
  */
-.button {
```
Comment (lyzadanger) [inline src/styles/sidebar/components/button.scss:12]: Before these changes, stuff was sort of randomly split between the `button` SCSS module and the `buttons` mixin. That has been rectified here. This alone was the source of a lot o' headache.

File: src/styles/sidebar/components/menu.scss [hunk 5]
```diff
@@ -23,6 +23,10 @@
     transition: 0.2s ease-out;
     color: var.$color-text;
   }
+  &[aria-expanded='true'],
```
Comment (lyzadanger) [inline src/styles/sidebar/components/menu.scss:26]: Attention: non-button change, technically. This is to help give on states to menu toggles.

File: src/styles/sidebar/components/search-status-bar.scss [hunk 6]
```diff
@@ -1,9 +1,13 @@
+@use '../../mixins/buttons';
+
 .search-status-bar {
   display: flex;
   align-items: center;
   margin-bottom: 10px;
 }
 
-.search-status-bar > button {
+.search-status-bar__button {
```
Comment (lyzadanger) [inline src/styles/sidebar/components/search-status-bar.scss:9]: Buh-bye specificity hack.

File: src/styles/sidebar/components/top-bar.scss [hunk 7]
```diff
@@ -18,8 +19,19 @@
   // the window is scrolled.
   transform: translate3d(0, 0, 0);
 
-  &__icon-button,
-  &__menu-icon {
+  &__icon-button {
+    @include buttons.button--icon-only;
+    padding: 0.25em;
+    @media (pointer: coarse) {
+      // Until the top bar can be refactored to allow for breathing room around
```
Comment (lyzadanger) [inline src/styles/sidebar/components/top-bar.scss:26]: This comment isn't new; it's been moved.

File: src/styles/sidebar/components/button.scss (hunk 4 above)
Comment (LMS007) [reply src/styles/sidebar/components/button.scss:12]: 👍

**Comment by lyzadanger** [comment]
@LMS007 

> general props
> I still think we may find a valid use for passing through generic props. One such case that comes to mind is the aria-haspopup for toggle buttons that active a menu. This of course breaks static linting and we would have to consider some work-arounds.

Hold that thought! Still valid, just outside of the scope of these changes.

**Comment by lyzadanger** [comment]
@LMS007 

> "__icon" class name
The second concern I have is with the implicitly ${baseClassName}__icon class. If we go with that, then we need to document that clearly. I'm personally in favor of just using decent selectors. I realize that breaks out class conventions so it's imperfect too.

Can you elaborate on "decent selectors?"

File: src/styles/mixins/buttons.scss [hunk 8]
```diff
@@ -67,3 +84,35 @@
     background-color: var.$grey-6;
   }
 }
+
+/**
+ * An icon-only button that sits to the right of a text-input field
+ * (e.g. "copy to clipboard" button in share panels)
+ */
+@mixin button--input {
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/styles/mixins/buttons.scss:92]: Quite possibly, but I think I'd vote to cross that bridge when we get there instead of presuming the pattern before it exists. What do you think?

File: src/sidebar/components/annotation-publish-control.js [hunk 9]
```diff
@@ -102,7 +102,7 @@ function AnnotationPublishControl({
         icon="cancel"
         buttonText="Cancel"
         onClick={onCancel}
-        useCompactStyle
+        className="annotation-publish-control__btn-cancel"
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/sidebar/components/annotation-publish-control.js:105]: Yes. It requires a class definition in an SCSS module, but the easier-to-use mixins I hope make it not too unpleasant.

I think the trade-off is worth it as it's very clear now (I hope!): 

* Pass a `className` to the `Button`? You're on your own for all styling.
* Don't pass a `className`? The `Button` will make a best attempt at default styling.

**Comment by lyzadanger** [comment]
@LMS007 I've pushed a few changes that add a little documentation to comments in `Button.js` (the component) and `buttons.scss` (the `buttons` SCSS mixin module). It was a bit of a balancing act to figure out where to add pointers without coupling the preact comments specifically to SCSS and vice-versa, but I hope this is in the right ballpark.

File: src/styles/mixins/buttons.scss (hunk 8 above)
[the remark this answers is not in the document]
Comment (LMS007) [reply src/styles/mixins/buttons.scss:111]: yep

File: src/styles/sidebar/components/annotation-publish-control.scss [hunk 10]
```diff
@@ -31,6 +32,16 @@
       min-width: 100%;
     }
 
+    &-cancel {
+      @include buttons.button--labeled;
+      padding-right: 0.5em;
+      padding-left: 0.5em;
+
+      &__icon {
```
Comment (LMS007) [inline src/styles/sidebar/components/annotation-publish-control.scss:40]: I don't see "__icon" bering added now, should this be
svg
or .svg-icon
?

File: src/sidebar/components/test/button-test.js [hunk 11]
```diff
@@ -115,28 +109,24 @@ describe('Button', () => {
...
 
-  it('adds additional class name passed in `className` prop', () => {
-    const wrapper = createComponent({ className: 'my-class' });
-
-    assert.isTrue(wrapper.hasClass('my-class'));
-  });
-
-  it('sets compact style if `useCompactStyle` is set`', () => {
-    const wrapper = createComponent({ useCompactStyle: true });
+  it('applies icon-only classname if no visible label present', () => {
```
Comment (LMS007) [inline src/sidebar/components/test/button-test.js:112]: both "class name" and "classname" both appear in this file. I suggest just calling it "class"

File: src/styles/mixins/buttons.scss [hunk 12]
```diff
@@ -37,28 +72,29 @@
...
-  padding-left: 0.75em;
+  font-weight: 700;
+  background-color: var.$grey-1;
 
   &:hover {
     background-color: var.$grey-2;
   }
+
+  svg,
+  &__icon {
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/styles/mixins/buttons.scss:87]: I was trying to account for the situation in which an icon might not be an `svg` component, or might not be the only descendant `svg` component in the `button`. That said, I suppose we could wait until such a situation actually arises...
